src/utils/bwellPersonFinder.js

```diff
@@ -0,0 +1,77 @@
+const {assertTypeEquals} = require('./assertType');
+const {DatabaseQueryFactory} = require('../dataLayer/databaseQueryFactory');
+
+class BwellPersonFinder {
+    /**
+     * constructor
+     * @param {DatabaseQueryFactory} databaseQueryFactory
+     */
+    constructor(
+        {
+            databaseQueryFactory
+        }
+    ) {
+        /**
+         * @type {DatabaseQueryFactory}
+         */
+        this.databaseQueryFactory = databaseQueryFactory;
+        assertTypeEquals(databaseQueryFactory, DatabaseQueryFactory);
+    }
+
+    /**
+     * finds the bwell person ID associated with a provided patient ID
+     * @param {string} patientId
+     * @return {Promise<string>}
+     */
+    async getBwellPersonIdAsync({ patientId}) {
+        const databaseQueryManager = this.databaseQueryFactory.createQuery({
+            resourceType: 'Person',
+            base_version: '4_0_0'
+        });
+
+        return await this.searchForBwellPersonAsync({
+            currentSubject: `Patient/${patientId}`,
+            databaseQueryManager: databaseQueryManager,
+            visitedSubjects: new Set()
+        });
+    }
+
+    /**
+     * recursively search through links to find a bwell Person
+     * @param {string} currentSubject
+     * @param {DatabaseQueryManager} databaseQueryManager for performing queries
+     * @param {Set} visitedSubjects subjects that have already been visited (to avoid infinite loops)
+     * @return {Promise<string>}
+     */
+    async searchForBwellPersonAsync({currentSubject, databaseQueryManager, visitedSubjects}) {
+        if (visitedSubjects.has(currentSubject)) {
+            return null;
+        }
+
+        visitedSubjects.add(currentSubject);
+
+        let foundPersonId = null;
+        let linkedPersons = await databaseQueryManager.findAsync({ query: { 'link.target.reference': currentSubject }});
+        while (!foundPersonId && linkedPersons && await linkedPersons.hasNext()) {
```

Comment:
Could this be an 'if' statement since it's recursing in the nested 'else' call?

Reply:
no, because the while loop is doing the breadth search and the recursion is doing the depth search. so there could be links to two different Persons and we'd want to call the recursive function on both of those. thus, a while loop over the two different Persons and the recursive call to go to the next layer of links

Reply:
I see, that makes sense.  Minor paranoid question, what are your thoughts on the runtime? I'm trying to trace through a run and what might happen if there are circular references to some of the non master Person records.

Reply:
I appreciate the paranoia. I did find a circular reference in the Person graph, and so that prompted the addition of this "visitedSubjects" Set. It's short-circuiting on line 47-48 whenever it finds one that it's already visited, which should protect us both from the cyclical references and the possibility of redundant (but non-cyclical) sub-searches. There are a couple unit tests to make sure that we don't get in an infinite loop if there's a cycle, whether or not there's actually a bwell person to be found in the graph. So I think we're sufficiently protected. But I share your paranoia.

Reply:
Ah nice! I see it now on line 50, we get the clean exit.  Ok, last paranoid question, should the 'foundPersonId' be outside the function? In the scenario that the person is found, but the other recursive calls are happening, we need to cancel those?

Reply:
Ah! good question. So for a given call to the bwellPersonFinder, I think the searches are running in serial. (i.e., we await on each result before starting the next one). consequently, it will only ever be checking one Person at a time. and once that's a bwell master person, it'll return all the way up the stack.
I'd be concerned about moving it outside of the function because (and this is based on some assumptions that may be wrong) this bwellPersonFinder may be used for multiple queries coming from different clients. if foundPersonId were a class variable instead of being scoped to a method, then those searches could cross the streams and interfere with each other. said another way, if one search got a foundPersonId and returned, then a new search came into the FHIR server and used the same bwellPersonFinder, it would just return that old foundPersonId result.
does that make sense? again, I'm new to node and making some assumptions about how this runs in practice

Reply:
Sounds good! Thanks for the clarification @justinpage-bwell

Reply:
@DarylCheung Thanks for your help!